compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -846,14 +847,23 @@ object TypeOps:
       var prefixTVar: Type | Null = null
       def apply(tp: Type): Type = tp match {
         case ThisType(tref: TypeRef) if !tref.symbol.isStaticOwner =>
+          val symbol = tref.symbol
           if (tref.symbol.is(Module))
             TermRef(this(tref.prefix), tref.symbol.sourceModule)
           else if (prefixTVar != null)
             this(tref)
           else {
             prefixTVar = WildcardType  // prevent recursive call from assigning it
-            val tref2 = this(tref.applyIfParameterized(tref.typeParams.map(_ => TypeBounds.empty)))
-            prefixTVar = newTypeVar(TypeBounds.upper(tref2))
+            prefixTVar = if symbol.is(Sealed) && symbol.isOneOf(AbstractOrTrait) && symbol.children.sizeIs > 0 && !symbol.hasAnonymousChild then
+              symbol.children.foldLeft(WildcardType: Type) { (acc, c) =>
+                val tp = this(c.appliedRef) match
+                  case tp: TypeRef if tp.symbol.is(ModuleVal) => tp.symbol.moduleClass.denot.thisType
+                  case tp => tp
+                acc | tp
+              }
```

Comment:
The solution is clever 👍 

For `this(c.appliedRef)`, I'm wondering if it can simply be `this(c)`? It's not clear what case the pattern match is handling, maybe add a comment. Also, a comment for the `if`-branch would be helpful.

In the first `if`-branch, it does not create type variables. What if the trait `RecordInstances` in the example is defined in another trait or contains type parameters?

Reply:
> In the first if-branch, it does not create type variables. What if the trait RecordInstances in the example is defined in another trait or contains type parameters?

I didn't follow this part?  I tried (locally) what I think you mean, and everything is still exhaustive as expected.  Could you share either what you think the code should do or the test case you think is problematic?

Reply:
Here is one example:

```Scala
// scalac: -Werror
sealed trait Schema[A]

sealed class Outer:
  sealed trait RecordInstances[B]:
    case class Field() extends Schema[B]
    case object Thing extends Schema[B]

object O1 extends Outer
object O2 extends Outer

object X extends O1.RecordInstances[Int]
object Y extends O2.RecordInstances[String]

// Match not exhaustive error! (with fatal warnings :P)
class Test:
  def handle(schema: Schema[String]) =
    schema match // was: match may not be exhaustive
      case Y.Field() =>
      case Y.Thing   =>
```

Reply:
Looks like that works as well.